File: baekjoon/graph/java/BOJ1987.java
import java.util.*;
import java.io.*;

class BOJ1987 {
    static int[][] graph;
    static int[] dr = {1, -1, 0, 0};
    static int[] dc = {0, 0, 1, -1};
    static int result = 0;
    static int R, C;
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        StringTokenizer st = new StringTokenizer(br.readLine());
        
        R = Integer.parseInt(st.nextToken());
        C = Integer.parseInt(st.nextToken());
        char offset = 'A';

        graph = new int[R][C];

        for (int i=0;i<R;i++) {
            String line = br.readLine();
            int j = 0;
            for (char ch : line.toCharArray()) {
                graph[i][j++] = ch - offset; 
            } 
        }
        

        dfs(0, 0, 1, 0);
        System.out.println(result);
    }

    public static void dfs(int r, int c, int count, int visited) {
        result = Math.max(result, count);

        visited |= (1 << graph[r][c]);

        for (int i=0;i<4;i++) {
            int nr = r + dr[i], nc = c + dc[i];

            if (nr >= 0 && nr < R && nc >= 0 && nc < C) {
                if ((visited & (1 << graph[nr][nc])) == 0) {
                    dfs(nr, nc, count + 1, visited);
                }
            }
        }
    }
}

File: baekjoon/graph/python/BOJ1987.py
from sys import stdin
readline = stdin.readline

R, C = map(int, readline().split())
graph = [readline().strip() for _ in range(R)]
dr, dc = [1, -1, 0, 0], [0, 0, 1, -1]

def solve():
    result = 1
    q = set()
    q.add((0, 0, graph[0][0]))

    while q and result < 26:
        row, col, string = q.pop()

        result = max(result, len(string))

        for i in range(4):
            nr, nc = row + dr[i], col + dc[i]

            if (0 <= nr < R and 0 <= nc < C and graph[nr][nc] not in string):
                q.add((nr, nc, string + graph[nr][nc]))

    print(result)            
solve()